class Solution {
    public boolean isValid(String s) {

        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < s.length(); i++) {
            char current = s.charAt(i);

            if (current == '(' || current == '[' || current == '{') {
                stack.push(current);
                continue;
            }

            if (stack.empty()) {
                return false;
            }

            char top = stack.pop();

            if (current == ')' && top != '(') {
                return false;
            }

            if (current == ']' && top != '[') {
                return false;
            }

            if (current == '}' && top != '{') {
                return false;
            }
        }

        return stack.empty();
    }
}
